pkg/kubelet/apis/kubeletconfig/types.go

```diff
@@ -40,15 +40,14 @@ const (
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// A configuration field should go in KubeletFlags instead of KubeletConfiguration if
-// its value cannot be safely shared between nodes at the same time (e.g. a hostname)
-// In general, please try to avoid adding flags or configuration fields,
-// we already have a confusingly large amount of them.
+// KubeletConfiguration contains the configuration for the Kubelet
 type KubeletConfiguration struct {
 	metav1.TypeMeta
 
-	// Only used for dynamic configuration.
-	// The length of the trial period for this configuration. This configuration will become the last-known-good after this duration.
+	// configTrialDuration is the length of the trial period for this configuration.
+	// This configuration will become the last-known-good after this duration.
+	// This field is only used for configurations delivered via dynamic kubelet configuration.
+	// This field is Alpha, and only takes effect when the DynamicKubeletConfig feature gate is enabled.
 	ConfigTrialDuration metav1.Duration
```

Comment:
if this is only used for dynamic config, doesn't it make more sense in the Node ConfigSource?

Reply:
I think this is a question of whether we should associate the measure of risk tolerance with the config being applied to a `Node` (`KubeletConfiguration.ConfigTrialDuration`), or with the `Node` the config is being applied to (`Node.ConfigSource.ConfigTrialDuration`). Now that you bring it up, I'm not sure I have a clear winner:

`KubeletConfiguration.ConfigTrialDuration`:
- \+ Directly associates idea of "risk" with a configuration, which makes some sense because whatever is *in* the configuration probably determines how risky it is.
- \- Feels weird because, as you mentioned, it only affects config from a specific delivery mechanism.
- +/- Simpler, in some ways, but less flexible (see below).

`Node.ConfigSource.ConfigTrialDuration`:
- \+ Can set the baseline for how "skeptical" a Node is of new configs, without having to determine the "risk" of each configuration.
- +/- Can technically achieve the same effect as `KubeletConfiguration.ConfigTrialDuration` by changing the assigned config and trial duration simultaneously, though you'd have to maintain that `(config, duration)` association independent of the config.
- \- Makes the `NodeConfigSource` API more complex, e.g. what does it mean to change the trial duration without changing the assigned config? I wonder if it would be confusing for a Kubelet to say: "Ah, config passed trial, now it's my last-known-good. Oh, you increased the trial? Guess now it's not my last known good anymore." This scenario requires the Kubelet to remember current and previous last-known-good configs, so that it can go back and forth when the trial duration changes. I worry this could be a confusing API (more implicit state and a parallel timeline).
- \+ Makes some sense when you think about higher-level declarative APIs that control config across pools of nodes; you probably want to allow these APIs to control config policy without having to edit serialized payloads.

As third option, we could consider `Node.ConfigSource.LastKnownGood`:
- +/- Elevates the problem of last-know-good selection to higher-level APIs. Inversion of control = inversion of responsibility.
- \+ Completely explicit.
- \- Kubelet is maybe less robust by default, though default fallback to "provisioned with" config is probably plenty good in circumstances where people have fungible VMs.

As a fourth option, we could remove the external knob altogether, have the Kubelet just use 10 minutes as an internal default, and decide where to expose control of last-known-good later on (I don't really like kicking the can down the road, but it's an option).

WDYT?

Reply:
I guess I have a hard time understanding how I would choose a value for this. It doesn't seem like something you'd tune super carefully... whatever gets you past the "I'm crash looping on config X, better roll back", which I'm not sure is best represented by a duration

Reply:
Without a clear picture of how we'd recommend someone tune this, I'd probably omit it and keep it internal and low(ish) for now, but happy to hear opposing views.

Reply:
I think that's probably what we should do for now, simply because it would be silly to block Kubelet's component config going beta while we argue this out.

Reply:
https://github.com/kubernetes/kubernetes/pull/59628